impl subscribe, unsubscribe, is-subscribed and double opt-in

siddhantk232 opened the pull request:
The behavior of these route handlers is defined in the README.md

siddhantk232 committed:
WIP: /subscribe

amitu commented on src/subscribe.rs:
```diff
@@ -0,0 +1,117 @@
+struct Subscriber {
+    name: Option<String>,
+    email: String,
+    phone: Option<String>,
```
Also add optional `source` and optional `topic`. In the luma import script you will find how we store `source` and `topic`.

siddhantk232 replied:
this struct holds subscriber info. The `source` and `topic` are added to the subscription below in the handler code.

amitu commented on src/subscribe.rs:
```diff
@@ -0,0 +1,117 @@
+
+impl Subscriber {
+    fn to_provider_data(&self) -> ft_sdk::auth::ProviderData {
+        ft_sdk::auth::ProviderData {
```
Subcription is not auth provider, it should not use ft_sdk::auth::ProviderData. The email stuff should be added a provider data, but for subscription (and any other app that wants to store user preferences etc) we should have another method.

siddhantk232 replied:
> If there is no user in the `fastn_user` table, this route creates one.

(from README.md:26)

you have to call `ft_sdk::auth::provider::create_user` with `ProviderData` to do this. FWIW, I am using "email" as the provider id to for this.

amitu commented on src/subscribe.rs:
```diff
@@ -0,0 +1,117 @@
+    ft_sdk::Query(email): ft_sdk::Query<"email", Option<String>>,
+    ft_sdk::Query(source): ft_sdk::Query<"source", Option<String>>,
+    ft_sdk::Query(topic): ft_sdk::Query<"topic", Option<String>>,
+    sid: ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
```
You will need another cookie,` fastn_tid`, instead of `fastn_sid`. We have to create `fastn_tracker` similar to `fastn_session`, but without `data`, as it will only keep track of user.

siddhantk232 replied:
creating a `tid` cookie is not the scope of this PR. However, the `is-subscribed` handler takes this cookie into account. The `/t/` handler will be added later via another PR

siddhantk232 committed:
impl /subscribe/

the `add_subscription_info` could be refactored but this is also fine
for now. I have not tested this yet

siddhantk232 committed:
impl /unsubscribe/

unsubscribe from all topics (set `subscribed: false`) if no `?topic` is
provided otherwise remove `?topic` from the subscriber's topic list.

WARN: not tested it yet, the code just compiles for now

siddhantk232 committed:
WIP setup double opt-in verification mail

modify the add subscription function to check if a new subscription is
added. A double opt-in email will be sent if this is true

siddhantk232 committed:
validate email input

siddhantk232 committed:
double opt-in in + /confirm-subscription/

- send a double opt-in email to the subscriber to confirm their
  subscription in `/subscribe/`

- a `confirm_subscription.rs` processor for verifying the subscription

siddhantk232 committed:
get user from session if `?email` is None...

...and move the handler to the top of the file

siddhantk232 committed:
impl /is-subscribed/

siddhantk232 committed:
manually create user + fix...

...fix: put source in sources. There's one case where it was put inside
topics

Arpita-Jaiswal commented on src/confirm_subscription.rs:
```diff
@@ -0,0 +1,74 @@
+        Err(e) => return Err(e.into()),
+    };
+
+    let data = {
+        let mut data = data;
+
+        data = mark_user_verified(data);
+
+        data
+    };
+    let data = serde_json::to_string(&data)?;
```
```suggestion
    let data = {
        let mut data = data;

        data = mark_user_verified(data);

        serde_json::to_string(&data)?
    };
```

Arpita-Jaiswal commented on src/confirm_subscription.rs:
```diff
@@ -0,0 +1,74 @@
+        data: String,
+    }
+
+    let (user_id, data): (i64, serde_json::Value)  = match diesel::sql_query(
+        r#"
+        SELECT id, data from fastn_user where json_extract(data, '$.subscription.confirmation_key') = $1 LIMIT 1;
+        "#,
+    )
```
Check if the `email` is also in data against the given `code`.

siddhantk232 replied:
Pretty sure this is redundant but done in https://github.com/fastn-community/subscription/pull/1/commits/10d1150656067ab794e70f79f7020f6fd5d1099a

Arpita-Jaiswal commented on src/confirm_subscription.rs:
```diff
@@ -0,0 +1,74 @@
+    };
+    let data = serde_json::to_string(&data)?;
+
+    diesel::update(
+        ft_sdk::auth::fastn_user::table.filter(ft_sdk::auth::fastn_user::id.eq(user_id)),
+    )
+    .set((
+        ft_sdk::auth::fastn_user::data.eq(&data),
+        ft_sdk::auth::fastn_user::updated_at.eq(ft_sdk::env::now()),
+    ))
+    .execute(&mut conn)?;
```
Use `ft_sdk::auth::provider::update_user`

siddhantk232 replied:
that can't be used here. Tha uses `HashMap::insert` on `data` which will not do deep merge. This is correct, for now.

amitu replied:
@siddhantk232 in that case we should use another method to update user from ft-sdk. We have to create an appropriate one, as others will also use it, and we can try to make sure ft-sdk upholds our assumptions.

Arpita-Jaiswal commented on src/confirm_subscription.rs:
```diff
@@ -0,0 +1,74 @@
+        data: String,
+    }
+
+    let (user_id, data): (i64, serde_json::Value)  = match diesel::sql_query(
+        r#"
+        SELECT id, data from fastn_user where json_extract(data, '$.subscription.confirmation_key') = $1 LIMIT 1;
+        "#,
+    )
```
@amitu I think we should use a similar structure as `ft_sdk::auth::ProviderData` so that we can use `ft_sdk::auth::provider::user_data_by_custom_attribute` etc to do read and write operations.

siddhantk232 replied:
how we store auth data is going to change definitely. We've agreed that the current way of storing everything as json is not working great for us. This will definitely change then but not now

Arpita-Jaiswal commented on src/is_subscribed.rs:
```diff
@@ -0,0 +1,107 @@
+
+fn check_if_subscribed(topic: Option<String>, user_data: serde_json::Value) -> bool {
+    let subscribed = if let Some(topic) = topic {
+        user_data
+            .as_object()
+            .map(|v| v.get("subscription"))
+            .flatten()
+            .map(|v| v.as_object())
+            .flatten()
+            .map(|v| v.get("topics"))
+            .flatten()
+            .map(|v| v.as_array())
+            .flatten()
+            .map(|v| v.iter().any(|t| t.as_str() == Some(&topic)))
+            .unwrap_or(false)
```
```suggestion
        user_data
            .as_object()
            .and_then(|v| v.get("subscription"))
            .and_then(|v| v.as_object())
            .and_then(|v| v.get("topics"))
            .and_then(|v| v.as_array())
            .map(|topics| topics.iter().any(|t| t.as_str() == Some(&topic)))
            .unwrap_or_default()
```

siddhantk232 replied:
thank you so much for this :heart:. I can never remember that `and_then` exists :sweat_smile:

Arpita-Jaiswal commented on src/is_subscribed.rs:
```diff
@@ -0,0 +1,107 @@
+            .map(|v| v.iter().any(|t| t.as_str() == Some(&topic)))
+            .unwrap_or(false)
+    } else {
+        user_data
+            .as_object()
+            .map(|v| v.get("subscription"))
+            .flatten()
+            .map(|v| v.as_object())
+            .flatten()
+            .map(|v| v.get("subscribed"))
+            .flatten()
+            .map(|v| v.as_bool())
+            .flatten()
+            .unwrap_or(false)
```
Similar to the above.

siddhantk232 replied:
done in https://github.com/fastn-community/subscription/pull/1/commits/575fdbd8eaf942f6c3fa884215d5de62ae1d7b32

Arpita-Jaiswal commented on src/is_subscribed.rs:
```diff
@@ -0,0 +1,107 @@
+    ft_sdk::Cookie(user_id_from_session): ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
+    ft_sdk::Cookie(user_id_from_tracker): ft_sdk::Cookie<"fastn-tid">,
+    conn: &mut ft_sdk::Connection,
+) -> Result<serde_json::Value, ft_sdk::Error> {
```
```suggestion
) -> Result<ft_sdk::AppData, ft_sdk::Error> {
```

ft_sdk::AppData somewhat similar to ft_sdk::auth::ProviderData

siddhantk232 replied:
not needed right now. An improvement for the future ;)

Arpita-Jaiswal commented on src/is_subscribed.rs:
```diff
@@ -0,0 +1,107 @@
+
+fn user_data_from_sid_or_tid(
+    ft_sdk::Cookie(user_id_from_session): ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
+    ft_sdk::Cookie(user_id_from_tracker): ft_sdk::Cookie<"fastn-tid">,
```
```suggestion
    ft_sdk::Cookie(user_id_from_tracker): ft_sdk::Cookie<ft_sdk::auth::TRACKER_KEY>,
```

siddhantk232 replied:
`TRACKER_KEY` does not exist. `TRACKER_KEY` is not an `auth` feature. It should come from `ft_sdk` though, I'll make this change when I actually implement setting tracker cookie. It's fine if it's hardcoded for now, it's only used in this one place.

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+    use diesel::prelude::*;
+    use ft_sdk::auth::fastn_user;
+
+    let data = serde_json::to_string(&serde_json::json!({
+        "email": {
+            "emails": [subscriber.email],
+            "identity": subscriber.email,
+        },
+        "subscription": {
+            "identity": subscriber.email,
+            "name": subscriber.name,
+        },
+    }))?;
```
This is subscription app. It should not add other provider data. Since we should use something like `ft_sdk::AppData` for all providers, other providers (like email auth) should take the leverage to get the required data.

Arpita-Jaiswal replied:
Also use `ft_sdk::auth::provider::create_user`

siddhantk232 replied:
We have not introduced the concept of `ft_sdk::AppData` yet. This code will change when that exists, we don't even know if we need that.

`ft_sdk::auth::provider::create_user` can't insert user data of this kind where multiple providers are set. This is wanted as the assumption that email should always be fetched from the `email` provider is made throughout the implementation of subscription and features in the [auth wasm](https://github.com/fastn-community/auth/) package

Arpita-Jaiswal commented on src/is_subscribed.rs:
```diff
@@ -0,0 +1,107 @@
+    let query;
+    if user_id_from_session.is_some() {
+        let user_id = user_id_from_session.unwrap();
+        query = diesel::sql_query(
+            r#"
+            SELECT
+                fastn_user.data as data
+            FROM fastn_user
+            JOIN fastn_session
+            WHERE
+                fastn_session.id = $1
+                AND fastn_user.id = fastn_session.uid
+            LIMIT 1;
+            "#,
+        )
```
I think we should never modify these tables from here. Instead, use ft-sdk functions. If the required function doesn't exist then add them. Since, in future, they'll be needed by other providers too.

siddhantk232 replied:
we will eventually have to expose functions that would just be wrapper around diesel, I don't feel that is required, we already give you the `ft_sdk::auth::fastn_user` table definition. We should give `fastn_session` and `fastn_tracker` from ft_sdk but that does not mean we start creating CRUD functions that are already pretty convenient made by `diesel`.

The custom sql query here could have been re-written using the diesel DSL. This, I'll fix later when we have definitions for `fastn_session` and `fastn_tracker` from somewhere.

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+    }
+
+    ft_sdk::println!("querying data");
+    let data: serde_json::Value = match diesel::sql_query(
+        r#"
+        SELECT data from fastn_user where id = $1 LIMIT 1;
+        "#,
+    )
+    .bind::<diesel::sql_types::Integer, _>(user_id.0)
+    .get_result::<UserData>(&mut conn)
+    {
+        Ok(d) => serde_json::from_str(&d.data)?,
+        Err(e) => return Err(e.into()),
+    };
```
We need subscription data only. Use `ft_sdk::auth::provider::user_data_by_id` to fetch and `ft_sdk::auth::provider::update_user` to update.

siddhantk232 replied:
> We need subscription data only

True. But most of the subscription data is incompatible with `ft_sdk::auth::ProviderData` returned by these `user_data_*` functions.

There are only a few keys that the subscription app interacts with (`subscribed`, `topics`, `sources`, `confirmed` etc) but there can be arbitrary keys here that may have been imported from other sources (benefits of schemaless). Mapping these keys to a struct that is nothing but just a bunch of `Option`s and then using `update` ensuring that the old values are preserved does not sound simpler or cleaner than the approach I've used here. I strongly feel that doing it your way would make this code any better. We can come back to this and improve it definitely in the future but I want to focus on getting the things that work to prod first.

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+    let (subscribed, mut data) = add_subscription_info(data, topic.clone(), source, &subscriber);
+
+    if subscribed {
+        let name = subscriber.name.unwrap_or_else(|| "".to_string());
```
```suggestion
        let name = subscriber.name.unwrap_or_default();
```

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+    (subscribed, data)
+}
+
+pub fn email_from_address_from_env() -> (String, String) {
+    let email = ft_sdk::env::var("FASTN_SMTP_SENDER_EMAIL".to_string())
+        .unwrap_or_else(|| "support@fifthtry.com".to_string());
+    let name = ft_sdk::env::var("FASTN_SMTP_SENDER_NAME".to_string())
+        .unwrap_or_else(|| "FifthTry Team".to_string());
+
+    (name, email)
+}
```
Let's move it to `ft-sdk`. Since it is used by `auth` also.

siddhantk232 replied:
Resolving this for now. This function will be removed in future when https://github.com/FifthTry/ft-sdk/pull/48 is merged :+1:

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+
+    let name_or_email = if to.0.is_empty() { to.1 } else { to.0 };
+
+    let to_topic = if topic.is_some() {
+        format!("to the {}", topic.unwrap())
+    } else {
+        "".to_string()
+    };
```
```suggestion
    let to_topic = topic.map(|topic| format!("to the {}", topic)).unwrap_or_default();
```

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+    let data = serde_json::to_string(&data)?;
+    ft_sdk::println!("querying data 3");
+
+    diesel::update(
+        ft_sdk::auth::fastn_user::table.filter(ft_sdk::auth::fastn_user::id.eq(user_id.0)),
+    )
+    .set((
+        ft_sdk::auth::fastn_user::data.eq(&data),
+        ft_sdk::auth::fastn_user::updated_at.eq(ft_sdk::env::now()),
+    ))
+    .execute(&mut conn)?;
```
Fix this to use `ft_sdk::auth::provider::update_user`

siddhantk232 replied:
I have explained why this can't happen today, read my replies above.

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+
+    ft_sdk::println!("querying data 4");
+
+    ft_sdk::form::redirect(next.unwrap_or_else(|| "/thank-you/".to_string()))
```
Mention in README.md that if using this wasm then there's a need to create a `/thank-you/` page. Also pass some query parameters to include some user data like name, topic etc so that the page is more personalised.

siddhantk232 replied:
already mentioned https://github.com/fastn-community/subscription/blob/main/README.md?plain=1#L28

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+    ft_sdk::Mountpoint(mountpoint): &ft_sdk::Mountpoint,
+) -> String {
+    format!(
+        "https://{host}{mountpoint}{confirm_sub_route}?code={key}&email={email}",
+        confirm_sub_route = "/confirm-subscription/",
```
```suggestion
        "https://{host}{mountpoint}/confirm-subscription/?code={key}&email={email}",
```

Arpita-Jaiswal commented on src/unsubscribe.rs:
```diff
@@ -0,0 +1,154 @@
+        data: String,
+    }
+
+    let data: serde_json::Value = match diesel::sql_query(
+        r#"
+        SELECT data from fastn_user where id = $1 LIMIT 1;
+        "#,
+    )
+    .bind::<diesel::sql_types::Integer, _>(user_id.0)
+    .get_result::<UserData>(&mut conn)
```
Do similar changes as mentioned for subscribe

Arpita-Jaiswal commented on src/subscribe.rs:
```diff
@@ -0,0 +1,429 @@
+) -> ft_sdk::form::Result {
+    use diesel::prelude::*;
+
+    let (subscriber, id) = validate(&mut conn, name, phone, email, sid)?;
+    let user_id = get_or_create_user_id(&mut conn, id, &subscriber)?;
```
Merge these two and do something similar to `unsubscribe::get_user`. It's much cleaner.

siddhantk232 replied:
No. Why? I don't see how rewriting this would make things clean

siddhantk232 committed:
minor refactor

Co-authored-by: Arpita Jaiswal <ayushipujaa@gmail.com>

siddhantk232 committed:
refactor: use `Option::and_then` instead of `map` and `flatten`

Co-authored-by: Arpita Jaiswal <ayushipujaa@gmail.com>

siddhantk232 committed:
no double opt-in for authenticated user + more...

- authenticated users don't get any email to confirm their subscription
- set `?next` to the link contained in the confirmation email. This
  value is taken from the `?on_confirm` in the `/subscribe/` handler.

siddhantk232 committed:
remove `println!` calls

siddhantk232 committed:
use `.and_then()` instead of `.map` then `.flatten`

siddhantk232 committed:
minor refactor

Co-authored-by: Arpita Jaiswal <ayushipujaa@gmail.com>

siddhantk232 committed:
confirm_subscription: add email check in the query

amitu commented on src/confirm_subscription.rs:
```diff
@@ -0,0 +1,83 @@
+        .get_mut("subscription")
+    {
+        Some(sub) => {
+            sub.as_object_mut()
+                .expect("subscription is always a json object")
+                .insert("confirmed".to_string(), serde_json::Value::Bool(true));
+
+            sub.as_object_mut()
+                .expect("subscription is always a json object")
+                .remove("confirmation_key");
```
Store result of `sub.as_object_mut().expect("subscription is always a json object”)` and do both operations on it.

siddhantk232 replied:
done in: https://github.com/fastn-community/subscription/pull/1/commits/3638c5d22780aa70dd131dd3e41d2eeb629da736

siddhantk232 committed:
refactor: reduce repeated operations

siddhantk232 committed:
send a welcome email after confirmed subscription
